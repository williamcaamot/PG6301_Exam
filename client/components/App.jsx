import React, {createContext, useEffect, useState} from "react";
import {BrowserRouter, Route, Router, Routes} from "react-router-dom";
import Login from "./login/Login.jsx";
import Header from "./Header.jsx";
import Home from "./home/Home.jsx";
import GoogleLoginCallback from "./login/GoogleLoginCallback.jsx";
import EntraIdLoginCallback from "./login/EntraIdLoginCallback.jsx";
import Chat from "./chat/Chat.jsx";
import Profile from "./profile/Profile.jsx";


export const AppContext = createContext();

export function App() {

    const [user, setUser] = useState(null);

    async function fetchUser() {
        const res = await fetch("/api/v1/login");
        if (res.status === 401) {
            setUser(undefined);

        } else {
            const user = await res.json();
            setUser(user);
            console.log(user)
        }
    }

    useEffect(() => {
        fetchUser();
    }, []);


    return <AppContext.Provider value={{user, setUser}}>
        <BrowserRouter>
        <Header/>
        <Routes>
            <Route path={"/login/callback/entraid"} element={<EntraIdLoginCallback/>}/>
            <Route path={"login/callback/google"} element={<GoogleLoginCallback/>}/>
            <Route path={"/login"} element={<Login/>}/>
            <Route path={"/profile"} element={<Profile/>}/>
            <Route path={"/chat"} element={<Chat/>}/>
            <Route path={"/"} element={<Home/>}/>
        </Routes>


    </BrowserRouter>
    </AppContext.Provider>
}