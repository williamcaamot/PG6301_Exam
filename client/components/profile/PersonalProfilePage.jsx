import React, {useContext, useState} from "react";
import {AppContext} from "../App.jsx";
import Profile from "./Profile.jsx";
import {Link, useNavigate} from "react-router-dom";

function PersonalProfilePage() {
    const [errorMessage, setErrorMessage] = useState();
    const [successMessage, setSuccessMessage] = useState();

    const navigate = useNavigate();
    const {user, setUser} = useContext(AppContext);

    const [chatRooms, setChatRooms] = useState();

    async function handleLogout() {
        try {
            await fetch("/api/v1/login", {
                method: "DELETE"
            });
            setUser(null);
            navigate("/");
        } catch (e) {
            setErrorMessage(e.message)
        }
    }

    async function loadChatRooms(){

    }


    return <>
        <div className={"pageContentWrapper"}>
            <div className={"innerWrapper"}>
                <Profile
                    user={user}/>
                <div style={{width: "100%", paddingTop:"20px"}}>
                    <button onClick={handleLogout}>Logg ut</button>
                    <Link to={"/profile/edit"}><button>Rediger profil</button></Link>

                </div>
                <h2>Dine chat rom:</h2>
            </div>

        </div>
    </>
}

export default PersonalProfilePage;