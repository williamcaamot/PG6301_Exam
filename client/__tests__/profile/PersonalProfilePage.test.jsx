import renderer, {act} from "react-test-renderer";
import {MemoryRouter} from "react-router-dom";
import PersonalProfilePage from "../../components/profile/PersonalProfilePage.jsx";
import {AppContext} from "../../components/App.jsx";

describe("render external profile page, the page other users can see", () => {

    it("should fetch profile details and dispaly them", async () => {

        const user = {
            _id: "655dd38a4c15f00c20bc3fcf",
            bio: "dawawddawJeg er kul",
            email: "williamcoucheronaamot@gmail.com",
            family_name: "Coucheron-Aamot",
            name: "William",
            nickname: "WILLIAM123wwwww",
            picture: "https://lh3.googleusercontent.com/a/ACg8ocIOgSmXkWTJqoKPe5Hk3JlkkhfUNer4sWkVo3UuIa_dzg=s96-c",
        }

        const chatRooms = [{
                _id: "655e00a9f595fb7c60770caa",
                description: "dawadwadwadw",
                messages: [],
                owner: "williamcoucheronaamot@gmail.com",
                title: "dawdwaadw"
            }, {
            _id: "655e02735da48dabbcbcb639",
                description: "daadwdwa",
                messages: [],
                owner: "williamcoucheronaamot@gmail.com",
                title: "dawdwaadw"
        }];

        async function setUser() {
        }
        let component;
        await act(async () => {
            component = renderer.create(
                <AppContext.Provider value={{user, setUser}}>
                    <MemoryRouter>
                        <PersonalProfilePage

                            handleLogout={() => {
                            }}

                            loadChatRooms={() => {
                                chatRooms
                            }}


                        />
                    </MemoryRouter>
                </AppContext.Provider>);
        });
        expect(component).toMatchSnapshot();
    })
})



