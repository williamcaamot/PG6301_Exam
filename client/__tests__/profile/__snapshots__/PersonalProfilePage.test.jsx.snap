// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render personal profile page should render profile details and chat rooms 1`] = `
<div
  className="pageContentWrapper"
>
  <div
    className="innerWrapper"
  >
    <div>
      <div
        className="userProfile"
      >
        <img
          alt="User Profile"
          src="https://lh3.googleusercontent.com/a/ACg8ocIOgSmXkWTJqoKPe5Hk3JlkkhfUNer4sWkVo3UuIa_dzg=s96-c"
        />
        <p>
          Kallenavn: 
          WILLIAM123wwwww
           
        </p>
        <p>
          Fult navn: 
          William
           
          Coucheron-Aamot
        </p>
        <p>
          E-post: 
          williamcoucheronaamot@gmail.com
        </p>
        <p>
          Om meg: 
          dawawddawJeg er kul
        </p>
      </div>
    </div>
    <div
      style={
        {
          "paddingTop": "20px",
          "width": "100%",
        }
      }
    >
      <button
        onClick={[Function]}
      >
        Logg ut
      </button>
      <a
        href="/profile/edit"
        onClick={[Function]}
      >
        <button
          id="editProfileButton"
        >
          Rediger profil
        </button>
      </a>
    </div>
    <h2>
      Dine chat rom:
    </h2>
  </div>
</div>
`;

exports[`render personal profile page should render profile then sign out 1`] = `
<div
  className="pageContentWrapper"
>
  <div
    className="innerWrapper"
  >
    <div>
      <div
        className="userProfile"
      >
        <img
          alt="User Profile"
          src="https://lh3.googleusercontent.com/a/ACg8ocIOgSmXkWTJqoKPe5Hk3JlkkhfUNer4sWkVo3UuIa_dzg=s96-c"
        />
        <p>
          Kallenavn: 
          WILLIAM123wwwww
           
        </p>
        <p>
          Fult navn: 
          William
           
          Coucheron-Aamot
        </p>
        <p>
          E-post: 
          williamcoucheronaamot@gmail.com
        </p>
        <p>
          Om meg: 
          dawawddawJeg er kul
        </p>
      </div>
    </div>
    <div
      style={
        {
          "paddingTop": "20px",
          "width": "100%",
        }
      }
    >
      <button
        onClick={[Function]}
      >
        Logg ut
      </button>
      <a
        href="/profile/edit"
        onClick={[Function]}
      >
        <button
          id="editProfileButton"
        >
          Rediger profil
        </button>
      </a>
    </div>
    <h2>
      Dine chat rom:
    </h2>
  </div>
</div>
`;
